p2p/transport/webrtc/stream_write.go

```diff
@@ -0,0 +1,204 @@
+package libp2pwebrtc
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"sync"
+	"time"
+
+	"github.com/libp2p/go-libp2p/p2p/transport/webrtc/internal/async"
+	pb "github.com/libp2p/go-libp2p/p2p/transport/webrtc/pb"
+	"github.com/libp2p/go-msgio/pbio"
+	"github.com/pion/datachannel"
+	"google.golang.org/protobuf/proto"
+)
+
+// Package pion detached data channel into a net.Conn
+// and then a network.MuxedStream
+type (
+	webRTCStreamWriter struct {
+		stream *webRTCStream
+
+		writer    pbio.Writer
+		writerMux sync.Mutex
+
+		deadline    time.Time
+		deadlineMux sync.Mutex
+
+		deadlineUpdated async.CondVar
+		writeAvailable  async.CondVar
+
+		readLoopOnce sync.Once
+		closeOnce    sync.Once
+	}
+)
+
+func (w *webRTCStreamWriter) Write(b []byte) (int, error) {
+	w.stream.wg.Add(1)
+	defer w.stream.wg.Done()
+
+	if !w.stream.stateHandler.AllowWrite() {
+		return 0, io.ErrClosedPipe
+	}
+
+	// Check if there is any message on the wire. This is used for control
+	// messages only when the read side of the stream is closed
+	if w.stream.stateHandler.State() == stateReadClosed {
+		w.readLoopOnce.Do(func() {
+			w.stream.wg.Add(1)
+			go func() {
+				defer w.stream.wg.Done()
+				// zero the read deadline, so read call only returns
+				// when the underlying datachannel closes or there is
+				// a message on the channel
+				w.stream.rwc.(*datachannel.DataChannel).SetReadDeadline(time.Time{})
+				var msg pb.Message
+				for {
+					if w.stream.stateHandler.Closed() {
+						return
+					}
+					err := w.stream.reader.state.readMessageFromDataChannel(&msg)
+					if err != nil {
+						if errors.Is(err, io.EOF) {
+							w.stream.close(true, true)
+						}
+						return
+					}
+					if msg.Flag != nil {
+						state, reset := w.stream.stateHandler.HandleInboundFlag(msg.GetFlag())
+						if state == stateClosed {
+							log.Debug("closing: after handle inbound flag")
+							w.stream.close(reset, true)
+						}
+					}
+				}
+			}()
+		})
+	}
+
+	const chunkSize = maxMessageSize - protoOverhead - varintOverhead
+
+	var (
+		err error
+		n   int
+	)
+
+	for len(b) > 0 {
+		end := len(b)
+		if chunkSize < end {
+			end = chunkSize
+		}
+
+		written, err := w.writeMessage(&pb.Message{Message: b[:end]})
+		n += written
+		if err != nil {
+			return n, err
+		}
+		b = b[end:]
+	}
+	return n, err
+}
+
+func (w *webRTCStreamWriter) writeMessage(msg *pb.Message) (int, error) {
```

Comment:
This signature implies that there are multiple different values that could be returned for the `int`, but it's always just 0 (for `err != nil`) or `len(msg.Message)` for `err == nil`. This suggests to remove the `int` return value altogether.